### PR TITLE
WIP: Add typescript typings

### PR DESCRIPTION
**Pull request by sergeysova**
Add more telegraf/telegram methods

Sort methods in Context and add more methods

Refactor Extras. Add replyWithAudio

Add chatAction

Rename all Extras. Add replyWithDocument

Add replyWithGame

Add replyWithHTML

Add replyWithInvoice

Add replyWithLocation

Add replyWithMarkdown

Add replyWithPhoto

Telegram types (#1)

* added some interfaces

* add InputFile interface

* described interface chatMember

* autoformatted

Add replyWithSticker

Add replyWithVideo

Add node typings and describe Interfaces (#2)

* added node typings

* described and added some interfaces

* refactored optional booleans in output interfaces

### CODE REVIEW

**Comment by sergeysova** [comment]
#71

**Comment by Y0hy0h** [comment]
I'm having some trouble using the branch. Can you maybe explain what work needs to be done and how I might be able to help? :)

**Comment by sergeysova** [comment]
@Y0hy0h Whats trouble do you have?

**Comment by Y0hy0h** [comment]
@sergeysova It's many small things, so it's hard to figure out for me what exactly is wrong.

But one example is this:
```typescript
import Telegraf from 'telegraf/typings'
...
async registerToBot(bot: Telegraf) {
    bot.start(async (ctx) => {if (ctx.chat.type === 'private') return this.childCommand(ctx)})
    bot.command('child', this.childCommand)
    bot.on('inline_query', ({inlineQuery, answerInlineQuery}) => this.inlineQuery(inlineQuery, answerInlineQuery))
    bot.on('chosen_inline_result', ({chosenInlineResult}) => this.chosenInlineResult(chosenInlineResult))
    bot.action('join', (ctx) => this.joinAction(ctx))
    bot.command('close', (ctx) => this.closeCommand(ctx))
}
```
WebStorm warns me of `bot.start` and `bot.action`, saying "Unresolved method". It also warns me that the parameters `this.childCommand` and `(ctx) => this.closeCommand(ctx)` are of the wrong type: 
For `this.childCommand` it warns that "Argument type (ctx: ContextMessageUpdate) => Promise<Promise<tt.Message>> is not assignable to parameter type Array<Middleware<ContextMessageUpdate>>`.

There are similar issues elsewhere in my code. You can [have a look at it](https://github.com/Y0hy0h/secret-santa-bot/tree/typescript) if you want to.

My problem is that I do not fully understand typings and even TypeScript for that matter. So I'm not sure what the problem is and how to fix it.

**Comment by elementh** [comment]
@sergeysova  I'm getting `TS2339: Property 'telegram' does not exist on type 'Telegraf<ContextMessageUpdate>'.` and `TS2339: Property 'options' does not exist on type 'Telegraf<ContextMessageUpdate>'.`

while trying to do this:

```javascript
bot.telegram.getMe().then((botInfo) => {
  bot.options.username = botInfo.username
})
```

Am I doing something wrong? 

Thanks!

**Comment by Piterden** [comment]
@elementh 
```js
bot.command('me', ({ replyWithMarkdown }) => {
  bot.telegram.getMe()
    .then((me) => {
      replyWithMarkdown('`' + JSON.stringify(me) + '`')
    })
})
```
This works fine ^

**Comment by sergeysova** [comment]
https://github.com/sergeysova/telegram-typings/blob/master/javascript/index.d.ts

**Comment by AndrewNeo** [comment]
The package.json seems to be missing the `types` field which is important if you intend on including typescript definitions in the npm package.

**Comment by oliver3** [comment]
As @AndrewNeo is saying, please add this to package.json :
```
  "types": "typings/index.d.ts",
```

File: .gitignore [hunk 1]
```diff
@@ -40,3 +40,4 @@ coverage
 _docpress
 package-lock.json
 .env
+.vscode
```
Comment (denisx) [inline .gitignore]: подобные исключение нужно держать в локальном юзер-глобале, а не таскать по репозиториям

**Comment by 1valdis** [comment]
Any progress on this? Can't wait for proper Intellisense.

**Pull request closed without merging**